### ISSUE TITLE
A dependacy on `path = "."` shuld have a good error mesege

### ISSUE THREAD

**Issue by Eh2406** [rust-lang/cargo#9518]
In a package `foo` a `[dev-dependencies] foo = { path = "." }` should be an error.
As there is no use case for it, and it does not do what most people expect.
There are not many people that will try this, and there are few configurations that don't error already. So I don't think we need to get too fancy with the comparison. A check for a path `"."` would be a big step foreword.

**Comment by Eh2406** [comment]
cc #9245 and https://github.com/rust-lang/cargo/pull/3921#issuecomment-297421636

**Comment by Nemo157** [comment]
Will probably fix https://github.com/rust-lang/cargo/issues/9454 too (by making it an error).

**Comment by yerke** [comment]
Does anyone have an example of a good error message we want to show in this case?

**Comment by Eh2406** [comment]
I don't have a good suggestion, but how does this sound:
"a project can not depend on itself as a path dependency"

**Comment by yerke** [comment]
@Eh2406 Great, thanks! It's better than "Something went wrong" I planned to use as a placeholder :)

**Comment by daxpedda** [comment]
This does have a good use case, it's a workaround for #2911 and works as intended as far as I've worked with it and tested it. I left a comment (https://github.com/rust-lang/cargo/issues/2911#issuecomment-898381422) explaining why, I think, there was a misconception around it not working as intended.

**Comment by davidhewitt** [comment]
@daxpedda - OTOH, we learned recently in PyO3 that using `path = .` dependency to enable features required for tests is brittle. `cargo` will strip the path dependency when packaging, so users who download the `.crate` from crates.io later will no longer be able to run `cargo test` without enabling all those features.

**Comment by daxpedda** [comment]
> so users who download the `.crate` from crates.io later will no longer be able to run `cargo test` without enabling all those features

Thanks for the heads up, I'm not really concerned though if tests work with `cargo test`. In any case, I agree that this "feature" should be removed eventually.

**Comment by davidhewitt** [comment]
TBH I reached the same conclusion in discussion on https://github.com/PyO3/pyo3/issues/1817

Having a way to specify features which are required for `cargo test` to work meaningfully is very convenient, and this self-dev-dependency trick definitely meets a need. I'd likewise be happy to replace it in PyO3 with something better, however that something better needs to be designed first...

**Comment by yerke** [comment]
Hi @daxpedda. You left a comment in https://github.com/rust-lang/cargo/pull/9702 saying that dependency on `.` should not be a hard error, but from your last comment in this issue I thought you agree that this should be removed.

@Eh2406 Do we have a decision on whether my PR is wanted or not?

**Comment by daxpedda** [comment]
I agree that it should **eventually** become an error, but currently it provides a workaround for a problem that has no other solution. Namely #2911.

Until there is a solution for this problem or a different workaround, I believe this should not become a hard error.

**Comment by davidhewitt** [comment]
We were able to remove the self-dev-dependency in PyO3 by using the `CARGO_PRIMARY_PACKAGE` env var, by using `if option_env!("CARGO_PRIMARY_PACKAGE").is_some()` to enable code paths only relevant for testing.

If `CARGO_PRIMARY_PACKAGE` were also set when executing `build.rs` then crates would be able to enable their own features / cfg based on this env var. I imagine this would be a more-than-sufficient alternative to https://github.com/rust-lang/cargo/issues/2911#issuecomment-749580481, thus enabling `path = "."` to become an error.

**Comment by Eh2406** [comment]
@yerke let me start off by apologizing. You deserve a prompt and knowledgeable response from me. This response is certainly not prompt and not as thorough as I would like it to be. In the time since I posted this issue, all the context has gone out of my head. I am doing my best to reconstruct it now.

For `dependencies` and `build-dependencies` a `path="."` already errors because it must create an infinite loop in the build order.
However things are more complicated for `dev-dependencies`. There always exists an implicit dev-dependency from the binarys/tests on the lib. The question is whether the `path="."` syntax unifies with that implicit edge or if it makes a new path package.
From my reading (but I have not gone back to test) different versions of Cargo have been inconsistent on the behavior of this syntax over time. Some versions it errored as building 2 incompatible versions of the same library, or it built it twice, or it unified.
Currently different aspects of cargo seem to have different opinions about what this means. (For example #9454, if you rename the `path="."` dependency it gets ignored and not marked as duplicate.) Generally tracking down these bugs and figuring out what is the correct behaviour is not worth the effort. Hence this issue to just ban the syntax.

On the other hand it happens to be that with `resolver="2"` the bugs line up to work ok as a workaround for #2911. Until a better work around is found (`CARGO_PRIMARY_PACKAGE`) or a permanent fix is implemented (https://github.com/rust-lang/rfcs/pull/3020) one can argue that we should let it be.

I will bring it up at the next Cargo Team meeting, and get back to you.

**Comment by yerke** [comment]
@Eh2406 Thank you very much for getting back to me and explaining the context.

**Comment by Eh2406** [comment]
I did bring it up at last week's Cargo meeting and then forgot to write the update here :facepalm: , Sorry.

> If `CARGO_PRIMARY_PACKAGE` were also set when executing `build.rs`

build.rs can set rust features, but it can not turn on optional dependencies. So this is also only a workaround.

We agreed that we do want to ban this syntax, eventually. Given that the most pressing bugs so far reported is #9454 (not very pressing) and that people are using it to work around #2911 we decided to not make it a hard error at this time. This decision can and will be re evaluated when better alternatives are available or we find more pressing bugs.

@yerke your PR is significantly more robust then the code I had in mind. Thank you for your contribution! When we ban this behavior, it will be based on your code. I am sorry I was slow and confusing in my communications, and that I left `E-help-wanted` on something it turned out we were not ready to do.

**Comment by jonaspleyer** [comment]
Sorry to bother again with this Issue. I am currently having the following problem: I have forked the `crates_io_api` crate and modified it such that I can use it in a WASM-specific crate. In order to use this package for publishing crates, I need to wait for my PR to be accepted and a new version to be published to crates.io.

In the meantime, I have created the `crates_io_api-wasm-patch` crate which is the same package but with the patch. For some of the tests, the full path `use crates_io_api::Something` is used.

I can create a dev-dependency like so:
```toml
[dev-dependencies]
crates_io_api = { version = "0.12.0", package = "crates_io_api-wasm-patch"}
```
Which brings all tests (including doctests!) to successfull completion. However, this approach will always test the version `0.12` on [crates.io](https://crates.io) rather than the local version. It is also not enough to use a git path here since testing would then always require pushing to the repository. Thus we need to specify something like this:

```toml
[dev-dependencies]
crates_io_api = { package = "crates_io_api-wasm-patch", path="./" }
```

This works for regular tests but not for doc-tests. Please let me know if there are other workarounds which I can use. I would like to avoid to rename all paths `crates_io_api::Something` in all files.

**Comment by weihanglo** [comment]
Have you tried [`[patch]`](https://doc.rust-lang.org/nightly/cargo/reference/overriding-dependencies.html#the-patch-section) and did it work?

**Comment by jonaspleyer** [comment]
> Have you tried [`[patch]`](https://doc.rust-lang.org/nightly/cargo/reference/overriding-dependencies.html#the-patch-section) and did it work?

I have just tried the following:
```toml
[dev-dependencies]
crates_io_api = { package = "crates_io_api-wasm-patch", git = "https://github.com/jonaspleyer/crates-io-api"}

[patch.'https://github.com/jonaspleyer/crates-io-api']
crates_io_api = { package = "crates_io_api-wasm-patch", path = "./" }
```

I inserted a `panic!()` inside the main `get` method which is used in (almost) every test. Most regular tests are failing but all doctests which also do contain this command are still running. So it seems to be equivalent to the `git = "..."` approach which I am currently using.

**Comment by Eh2406** [comment]
I think you want something more like:
```toml
[dev-dependencies]
crates_io_api = { version = "0.12.0", package = "crates_io_api-wasm-patch"}

[patch.crates-io]
crates_io_api-wasm-patch = { path = "./" }
```
Specifically, the `[dev-dependencies]` stays looking exactly like it's pulling from `crates-io`. The `patch` section says that it should instead come from `path = "./"`.

However, none of this has to do with this issue for adding an error message to Cargo. If you need more help using the patch section please try one of the user forums like https://users.rust-lang.org/
